server/interceptor.go

```diff
@@ -0,0 +1,140 @@
+)
+
+type interceptor struct {
+	jobUserStore store.JobUserStore
+}
+
+func NewInterceptor(store store.JobUserStore) *interceptor {
+	return &interceptor{
+		jobUserStore: store,
+	}
+}
+
+// UnaryAuthInterceptor intercept unary calls to authorize the user
+// It checks user role using certification extension oid 1.2.840.10070.8.1, it also checks if user has access to the requested resource
+func (i *interceptor) UnaryAuthInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	jobID := interceptRequest(info.FullMethod, req)
+	newCtx, err := i.authorize(ctx, info.FullMethod, jobID)
+	if err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+	return handler(newCtx, req)
+}
+
+type recvWrapper struct {
+	grpc.ServerStream
+	ctx context.Context
+	*interceptor
+}
+
+func (r *recvWrapper) RecvMsg(m interface{}) error {
+	if err := r.ServerStream.RecvMsg(m); err != nil {
+		logrus.Errorf("failed to intercept stream: %v", err)
+		return err
+	}
+	if req, ok := m.(*proto.GetStreamRequest); ok {
+		jobID := req.GetId()
+		newCtx, err := r.authorize(r.ctx, "/proto.WorkerService/GetJobStream", jobID)
+		if err != nil {
+			return err
+		}
+		md, _ := metadata.FromIncomingContext(newCtx)
+		r.ctx = metadata.NewIncomingContext(newCtx, md)
+	}
+	return nil
+}
+
+// StreamAuthInterceptor intercept stream calls to authorize the user
+// It checks the user role using certification extension oid 1.2.840.10070.8.1, it also checks if user has access to the requested resource
+func (i *interceptor) StreamAuthInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	wrapper := &recvWrapper{stream, stream.Context(), i}
+	return handler(srv, wrapper)
+}
+
+// authorize verifies the user information given by certificate
+// against the mapped roles for a specific method and also checks if user has access to requested job
+func (i *interceptor) authorize(ctx context.Context, method string, jobID string) (context.Context, error) {
+	// reads the peer information from context
+	peer, ok := peer.FromContext(ctx)
+	if !ok {
+		return ctx, fmt.Errorf("error to read peer information")
+	}
+	// reads user tls information
+	tlsInfo, ok := peer.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return ctx, fmt.Errorf("error to get auth information")
+	}
+	// get user roles
+	certs := tlsInfo.State.VerifiedChains
+	if len(certs) == 0 || len(certs[0]) == 0 {
+		return ctx, fmt.Errorf("missing certificate chain")
+	}
+
+	// find user roles from certificate extensions
+	var roles []string
+	for _, ext := range certs[0][0].Extensions {
+		if oid := OidToString(ext.Id); IsOidRole(oid) {
+			roles = ParseRoles(string(ext.Value))
+			break
+		}
+	}
+	// check user has access to execute a specific method
+	if !HasAccess(method, roles) {
+		return ctx, fmt.Errorf("unauthorized, user does not have privileges")
```

Comment:
When you don't have any format verbs, it's better to use `errors.New` Same applies in several places below.

```suggestion
		return ctx, errors.New("unauthorized, user does not have privileges")
```